service/storage_proxy.cc

```diff
@@ -4676,12 +4680,25 @@ void query_ranges_to_vnodes_generator::process_one_range(size_t n, dht::partitio
         return;
     }
 
+    auto is_empty_range = [] (const std::optional<dht::token> s, const std::optional<dht::token> e) -> bool {
+        bool case_1 = e && e->is_minimum();
+        bool case_2 = s && s->is_maximum();
+        return case_1 || case_2;
+    };
+    auto is_singular_range = [] (const std::optional<dht::token> s, const std::optional<dht::token> e) -> bool {
+        return s && e && *s == *e;
+    };
+
+    std::optional<dht::token> start = start_token(cr);
+    std::optional<dht::token> end = end_token(cr);
+
+    if (is_empty_range(start, end)) {
+        _i++; // empty range? Move to the next one
+        return;
+    }
+
     // special case for bounds containing exactly 1 token
-    if (start_token(cr) == end_token(cr)) {
-        if (start_token(cr).is_minimum()) {
-            _i++; // empty range? Move to the next one
-            return;
-        }
+    if (is_singular_range(start, end)) {
```

Comment:
Perhaps this special case can be removed. But let's not mix it with this change.